fenix/app/src/main/java/org/mozilla/fenix/library/history/state/HistorySyncMiddleware.kt

```diff
@@ -0,0 +1,57 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.library.history.state
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.service.fxa.SyncEngine
+import mozilla.components.service.fxa.manager.FxaAccountManager
+import mozilla.components.service.fxa.sync.SyncReason
+import org.mozilla.fenix.FeatureFlags
+import org.mozilla.fenix.library.history.HistoryFragmentAction
+import org.mozilla.fenix.library.history.HistoryFragmentState
+
+/**
+ * A [Middleware] for handling Sync side-effects based on [HistoryFragmentAction]s that are dispatched
+ * to the [HistoryFragmentStore].
+ *
+ * @property accountManager The [FxaAccountManager] for handling Sync operations.
+ * @property refreshView Callback to refresh the view once a Sync is completed.
+ * @property scope Coroutine scope to launch Sync operations into.
+ */
+class HistorySyncMiddleware(
+    private val accountManager: FxaAccountManager,
+    private val refreshView: () -> Unit,
+    private val scope: CoroutineScope,
+) : Middleware<HistoryFragmentState, HistoryFragmentAction> {
+    override fun invoke(
+        context: MiddlewareContext<HistoryFragmentState, HistoryFragmentAction>,
+        next: (HistoryFragmentAction) -> Unit,
+        action: HistoryFragmentAction,
+    ) {
+        next(action)
+        if (!FeatureFlags.historyFragmentLibStateRefactor) {
+            // StartSync is also fired by the interactor, so we should avoid syncing twice in cases
+            // where it is still being used.
+            return
+        }
```

Comment:
nit thought: alternatively, you could put the feature flag check in the store's middleware intialization

Reply:
Oh, good call!